raft/confchange/confchange.go

```diff
@@ -0,0 +1,375 @@
+// refusing invalid configuration changes before they affect the active
+// configuration.
+type Changer struct {
+	Tracker   tracker.ProgressTracker
+	LastIndex uint64
+}
+
+// EnterJoint verifies that the right half of the joint quorum is empty. Then it
+// copies the left half into the right half. That is, it goes from
+//
+//     (1 2 3)
+// to
+//     (1 2 3)&&(1 2 3)
+//
+// It then applies the supplied ConfChanges to the left half of the
+// configuration to arrive at the desired joint configuration.
+func (c Changer) EnterJoint(ccs ...pb.ConfChange) (tracker.Config, tracker.ProgressMap, error) {
+	cfg, prs, err := c.checkAndCopy()
+	if err != nil {
+		return c.err(err)
+	}
+	if joint(cfg) {
+		err := errors.New("config is already joint")
+		return c.err(err)
+	}
+	if len(cfg.Voters[0]) == 0 {
+		err := errors.New("can't make a zero-voter config joint")
+		return c.err(err)
+	}
+	cfg.Voters[1] = quorum.MajorityConfig{}
+	for id := range cfg.Voters[0] {
+		cfg.Voters[1][id] = struct{}{}
+	}
+
+	if err := c.apply(&cfg, prs, ccs...); err != nil {
+		return c.err(err)
+	}
+
+	return cfg, prs, nil
+}
+
+// LeaveJoint transitions out of a joint configuration. It is an error to call
+// this method if the configuration is not joint, i.e. if Voters[1] is empty.
+//
+// The right side of the joint configuration will be removed, and any staged
+// learners (LearnersNext) will be inserted into Learners.
+func (c Changer) LeaveJoint() (tracker.Config, tracker.ProgressMap, error) {
+	cfg, prs, err := c.checkAndCopy()
+	if err != nil {
+		return c.err(err)
+	}
+	if !joint(cfg) {
+		err := errors.New("can't leave a non-joint config")
+		return c.err(err)
+	}
+	if len(cfg.Voters[1]) == 0 {
+		err := fmt.Errorf("configuration is not joint: %v", cfg)
+		return c.err(err)
+	}
+	for id := range cfg.LearnersNext {
+		nilAwareAdd(&cfg.Learners, id)
+		prs[id].IsLearner = true
+	}
+	cfg.LearnersNext = nil
+
+	for id := range cfg.Voters[1] {
+		_, isVoter := cfg.Voters[0][id]
+		_, isLearner := cfg.Learners[id]
+
+		if !isVoter && !isLearner {
+			delete(prs, id)
+		}
+	}
+	cfg.Voters[1] = nil
+	return cfg, prs, nil
+}
+
+// Simple carries out a series of configuration changes that (in aggregate)
+// mutates the set of voters by at most one. It will return an error if that is
+// not the case or if the resulting quorum is zero.
+func (c Changer) Simple(ccs ...pb.ConfChange) (tracker.Config, tracker.ProgressMap, error) {
+	cfg, prs, err := c.checkAndCopy()
+	if err != nil {
+		return c.err(err)
+	}
+	if joint(cfg) {
+		err := errors.New("can't apply simple config change in joint config")
+		return c.err(err)
+	}
+	if err := c.apply(&cfg, prs, ccs...); err != nil {
+		return c.err(err)
+	}
+	if n := symdiff(c.Tracker.Voters[0], cfg.Voters[0]); n > 1 {
+		return tracker.Config{}, nil, errors.New("more than voter changed without entering joint config")
+	}
+	return cfg, prs, nil
+}
+
+func (c Changer) next() uint64 {
```

Comment:
It's not clear what the point of this method is outside of the one place that it's used. Should it be inlined?